include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+    {
+      const auto end_mark = new_pos - 1;
+      auto dst = end() + elements_to_insert - 1;
+      auto src = dst - elements_to_insert;
+      while (src != end_mark) {
+        allocator_.construct(dst, std::move(*src));
+        allocator_.destroy(src); // destruct residual object
+        --src;
+        --dst;
+      }
+    }
+    // copy construct the new contents in place. The places are currently empty
+    {
+      auto dst = new_pos;
+      for (auto it = first; it != last; ++it, ++dst) {
+        allocator_.construct(dst, *it);
+      }
+    }
+    size_ += elements_to_insert;
+  }
+  return new_pos;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::insert(const_iterator pos,
+                                   std::initializer_list<Type> ilist) {
+  return insert(pos, ilist.begin(), ilist.end());
+}
+
+template <class Type, class Allocator>
+template <class... Args>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::emplace(const_iterator pos, Args &&... args) {
+  const auto pos_ordinal = empty() ? 0 : std::distance(cbegin(), pos);
+  emplace_back(std::forward<Args>(args)...);
+  const auto new_pos = begin() + pos_ordinal;
+  auto index = end() - 1;
+  while (index != new_pos) {
+    std::swap(*index, *(index - 1));
+    --index;
+  }
+  return new_pos;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::erase(const_iterator pos) {
+  /* Move the element at pos towards the end of the vector, then remove it */
+  iterator non_const_pos = begin() + std::distance(cbegin(), pos);
+  auto index = non_const_pos;
+  while ((index + 1) != end()) {
+    std::swap(*index, *(index + 1));
+    ++index;
+  }
+  pop_back();
+  return non_const_pos;
+}
+
+template <class Type, class Allocator>
+typename ekuvector<Type, Allocator>::iterator
+ekuvector<Type, Allocator>::erase(const_iterator first, const_iterator last) {
+  /* Move the elements within both iterators towards the end, then remove them */
+  auto head = begin() + std::distance(cbegin(), first);
+  auto tail = begin() + std::distance(cbegin(), last);
+  auto return_it = head;
+
+  while (tail != end()) {
+    std::swap(*head, *tail);
+    ++head;
+    ++tail;
+  }
+
+  while (head != end()) {
+    pop_back();
+  }
+
+  return return_it;
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::push_back(const Type &value) {
+  /* make sure there's enough storage */
+  preallocate_capacity(size_ + 1);
+  /* copy construct the new element */
+  allocator_.construct(data_ + size_, value);
+  ++size_;
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::push_back(Type &&value) {
+  /* make sure there's enough storage */
+  preallocate_capacity(size_ + 1);
+  /* move construct the new element */
+  allocator_.construct(data_ + size_, std::move(value));
```

Comment:
@glpuga no need to `std::move`, it is already an rvalue reference.

Reply:
Good catch. Fixed.